### DISCUSSION TITLE
Isomorphism with MultiDiGraph and weighted edges

### DISCUSSION THREAD

**Discussion by npbauman** [networkx/networkx#7409] [Q&A]
I am trying to get NetworkX to determine if two graphs are equivalent. Some things about the graphs:

1. There are 5 vertices labeled "A", "B" (twice), "C", and "D".
2. Edges to and from vertex "A" are given weights that do not repeat.
3. The remaining edges do not have weights.

The following is an example of two graphs I am comparing:
```import networkx as nx

G1 = nx.MultiDiGraph()
G2 = nx.MultiDiGraph()
G1.add_node(0, name = "A")
G1.add_node(1, name = "D")
G1.add_node(2, name = "B")
G1.add_node(3, name = "B")
G1.add_node(4, name = "C")
G2.add_node(0, name = "A")
G2.add_node(1, name = "D")
G2.add_node(2, name = "B")
G2.add_node(3, name = "B")
G2.add_node(4, name = "C")
G1.add_edges_from([(0, 4, {'weight': 1}), (4, 0, {'weight': 4}), (1, 4), (4, 1), (1, 3), (2, 1), (0, 2, {'weight': 2}), (3, 0, {'weight': 3})])
G2.add_edges_from([(0, 4, {'weight': 2}), (4, 0, {'weight': 4}), (1, 4), (4, 1), (1, 3), (2, 1), (0, 2, {'weight': 1}), (3, 0, {'weight': 3})])
nm = nx.isomorphism.categorical_node_match(["name","name","name","name"], ["A","B","C","D"])
em = nx.isomorphism.numerical_edge_match(['weight','weight','weight','weight'], [1,2,3,4])
print(nx.is_isomorphic(G1,G2, node_match=nm, edge_match=em))
```

The is_isomorphic function returns True for this case. However, I thought the "weight" attribute would make the function return False because the edges with weights 1 and 2 go to different vertices.

Is this expected, and if so, does it make it such that these two graphs return False?

The graphs can be seen here:
https://stackoverflow.com/questions/78313307/isomorphism-with-multidigraph-and-weighted-edges

**Comment by dschult** [comment]
As you are using a MultiDiGraph, you should use `numerical_multiedge_match` instead of `numerical_edge_match`. 
Then the inputs to that function don't match how you are constructing it. You have one node attribute. So to match based on it you should provide a single attribute name, and a single default value (for nodes that don't have any attribute with that name.  Something like: `em = nx.isomorphism.categorical_node_match('name', None)`.
Similarly, you only have one edge attribute, so:  `em = nx.isomorphism.numerical_multiedge_match("weight", 0.5)`  (or whatever value you want the edges to have when they don't have an edge weight named "weight"). 

You can see what the resulting edge match function actually is in ipython or jupyter using `em??` which shows you the code for the function. The docs for the `is_isomorphic` function explain how to make an `edge_match` function for the `MultiDiGraph` case. 

```python
>>> nm = nx.isomorphism.categorical_node_match('name', None)
>>> em = nx.isomorphism.numerical_multiedge_match("weight", 0.5)
>>> nx.is_isomorphic(G1, G2)
True
>>> nx.is_isomorphic(G1, G2, node_match=nm)
True
>>> nx.is_isomorphic(G1, G2, node_match=nm, edge_match=em)
False
```

**Reply by npbauman** [reply]
This was greatly helpful! Thank you very much. I have been reading and looking at examples but have never fully understood how to use the attributes to match until now.

**Reply by dschult** [reply]
Good -- it looks like we could improve the docs then. If you have suggestions in that direction, let us know. :)
